File: data_agent/super_agent/diagnostics_sanitizer.py
"""Generic separation of internal execution telemetry from business-facing conclusions."""

from __future__ import annotations

import re
from collections.abc import Iterable

_INTERNAL_TELEMETRY_KEYS = (
    "execution_mode_summary",
    "scheduler_summary",
    "task_board_summary",
    "fallback_reason",
    "harness_count",
    "generic_llm_harness_count",
    "deterministic_count",
    "blocked_count",
    "failed_count",
)

_TELEMETRY_PATTERNS: tuple[re.Pattern[str], ...] = (
    re.compile(r"\bexecution_mode_summary\s*[=:]\s*[\{\[]", re.I),
    re.compile(r"\bscheduler_summary\s*[=:]\s*[\{\[]", re.I),
    re.compile(r"\btask_board_summary\s*[=:]\s*[\{\[]", re.I),
    re.compile(r"\bfallback_reason\s*=", re.I),
    re.compile(r"\b\w+_summary\s*=\s*[\{\[]"),
    re.compile(r"\bexecution_mode\s*=\s*[\{]", re.I),
    re.compile(r"\bexecution_mode\s*=\s*deterministic", re.I),
    re.compile(r"\blimited\s*=\s*(true|false)", re.I),
    re.compile(r"\bbootstrap_mode\s*=", re.I),
    re.compile(r"\bgate_limited:", re.I),
    re.compile(
        r"\b(harness_count|generic_llm_harness_count|deterministic_count|blocked_count|failed_count)\s*[=:]\s*\d",
        re.I,
    ),
    re.compile(
        r"['\"](harness_count|generic_llm_harness_count|deterministic_count|blocked_count|failed_count)['\"]\s*:\s*\d",
        re.I,
    ),
    re.compile(r"\btraceability_summary\s*=", re.I),
    re.compile(r"\bdeterministic_items\s*=", re.I),
)

_DICT_TELEMETRY_INNER = re.compile(
    r"[\{\[][^}\]]*"
    r"(harness_count|generic_llm_harness_count|deterministic_count|blocked_count|failed_count|execution_mode)"
    r"[^}\]]*[\}\]]",
    re.I,
)

_KEY_VALUE_DICT_LINE = re.compile(
    r"^[\w.\s/-]+=\{.*\}$",
)


def _cjk_char_count(text: str) -> int:
    return sum(1 for char in text if "\u4e00" <= char <= "\u9fff")


def is_internal_diagnostic_text(text: str) -> bool:
    """Return True when *text* is execution telemetry rather than a business conclusion."""
    normalized = str(text or "").strip()
    if not normalized:
        return False

    for pattern in _TELEMETRY_PATTERNS:
        if pattern.search(normalized):
            return True

    if _DICT_TELEMETRY_INNER.search(normalized):
        return True

    if _KEY_VALUE_DICT_LINE.match(normalized):
        inner = normalized.split("=", 1)[1]
        if any(key in inner for key in _INTERNAL_TELEMETRY_KEYS):
            return True

    lowered = normalized.lower()
    if lowered.startswith("smart committee limited:") and "{" in normalized:
        return True

    # Mostly Chinese prose without inline dict/json payloads — treat as business text.
    if _cjk_char_count(normalized) >= max(4, len(normalized) // 3):
        if not re.search(r"[\{\[]", normalized):
            return False
        if not any(key in normalized for key in _INTERNAL_TELEMETRY_KEYS):
            return False

    return False


def sanitize_business_lines(lines: Iterable[str]) -> list[str]:
    """Drop internal telemetry lines while preserving order and deduplicating."""
    seen: set[str] = set()
    filtered: list[str] = []
    for item in lines:
        text = str(item or "").strip()
        if not text or is_internal_diagnostic_text(text):
            continue
        if text in seen:
            continue
        seen.add(text)
        filtered.append(text)
    return filtered


def sanitize_report_markdown(markdown: str) -> str:
    """Remove internal diagnostic bullet/paragraph lines from report markdown."""
    if not markdown:
        return ""
    kept: list[str] = []
    for line in markdown.splitlines():
        stripped = line.lstrip()
        if stripped.startswith("- "):
            content = stripped[2:].strip()
            if is_internal_diagnostic_text(content):
                continue
        elif stripped.startswith("* "):
            content = stripped[2:].strip()
            if is_internal_diagnostic_text(content):
                continue
        elif is_internal_diagnostic_text(line.strip()):
            continue
        kept.append(line)
    return "\n".join(kept)


# User-facing aliases
sanitize_smart_diagnostic_text = sanitize_report_markdown
sanitize_business_report_text = sanitize_report_markdown

__all__ = [
    "is_internal_diagnostic_text",
    "sanitize_business_lines",
    "sanitize_business_report_text",
    "sanitize_report_markdown",
    "sanitize_smart_diagnostic_text",
]

File: web/src/features/super-agent/utils/diagnosticsSanitizer.ts
const INTERNAL_TELEMETRY_KEYS = [
  'execution_mode_summary',
  'scheduler_summary',
  'task_board_summary',
  'fallback_reason',
  'harness_count',
  'generic_llm_harness_count',
  'deterministic_count',
  'blocked_count',
  'failed_count',
] as const

const TELEMETRY_PATTERNS: RegExp[] = [
  /\bexecution_mode_summary\s*[=:]\s*[\{\[]/i,
  /\bscheduler_summary\s*[=:]\s*[\{\[]/i,
  /\btask_board_summary\s*[=:]\s*[\{\[]/i,
  /\bfallback_reason\s*=/i,
  /\b\w+_summary\s*=\s*[\{\[]/,
  /\bexecution_mode\s*=\s*[\{]/i,
  /\bexecution_mode\s*=\s*deterministic/i,
  /\blimited\s*=\s*(true|false)/i,
  /\bbootstrap_mode\s*=/i,
  /\bgate_limited:/i,
  /\b(harness_count|generic_llm_harness_count|deterministic_count|blocked_count|failed_count)\s*[=:]\s*\d/i,
  /['"](harness_count|generic_llm_harness_count|deterministic_count|blocked_count|failed_count)['"]\s*:\s*\d/i,
  /\btraceability_summary\s*=/i,
  /\bdeterministic_items\s*=/i,
]

const DICT_TELEMETRY_INNER = /[\{\[][^}\]]*(harness_count|generic_llm_harness_count|deterministic_count|blocked_count|failed_count|execution_mode)[^}\]]*[\}\]]/i

const KEY_VALUE_DICT_LINE = /^[\w.\s/-]+=\{.*\}$/

function cjkCharCount(text: string): number {
  let count = 0
  for (const char of text) {
    const code = char.codePointAt(0) || 0
    if (code >= 0x4e00 && code <= 0x9fff) count += 1
  }
  return count
}

export function isInternalDiagnosticText(text: string): boolean {
  const normalized = String(text || '').trim()
  if (!normalized) return false

  if (TELEMETRY_PATTERNS.some((pattern) => pattern.test(normalized))) {
    return true
  }

  if (DICT_TELEMETRY_INNER.test(normalized)) {
    return true
  }

  if (KEY_VALUE_DICT_LINE.test(normalized)) {
    const inner = normalized.split('=', 2)[1] || ''
    if (INTERNAL_TELEMETRY_KEYS.some((key) => inner.includes(key))) {
      return true
    }
  }

  const lowered = normalized.toLowerCase()
  if (lowered.startsWith('smart committee limited:') && normalized.includes('{')) {
    return true
  }

  if (cjkCharCount(normalized) >= Math.max(4, Math.floor(normalized.length / 3))) {
    if (!/[\{\[]/.test(normalized)) return false
    if (!INTERNAL_TELEMETRY_KEYS.some((key) => normalized.includes(key))) return false
  }

  return false
}

export function filterBusinessLines(lines: string[]): string[] {
  const seen = new Set<string>()
  const filtered: string[] = []
  for (const item of lines) {
    const text = String(item || '').trim()
    if (!text || isInternalDiagnosticText(text) || seen.has(text)) {
      continue
    }
    seen.add(text)
    filtered.push(text)
  }
  return filtered
}

export function sanitizeBusinessMarkdown(markdown: string): string {
  if (!markdown) return ''
  const kept: string[] = []
  for (const line of markdown.split('\n')) {
    const stripped = line.trimStart()
    if (stripped.startsWith('- ')) {
      const content = stripped.slice(2).trim()
      if (isInternalDiagnosticText(content)) continue
    } else if (stripped.startsWith('* ')) {
      const content = stripped.slice(2).trim()
      if (isInternalDiagnosticText(content)) continue
    } else if (isInternalDiagnosticText(stripped)) {
      continue
    }
    kept.push(line)
  }
  return kept.join('\n')
}

export function sanitizeSmartDiagnosticText(text: string): string {
  return sanitizeBusinessMarkdown(text)
}

export function sanitizeBusinessReportText(text: string): string {
  return sanitizeBusinessMarkdown(text)
}

/** @deprecated Use isInternalDiagnosticText */
export const isSmartInternalDiagnostic = isInternalDiagnosticText

/** @deprecated Use filterBusinessLines */
export const filterBusinessFindings = filterBusinessLines
